src/bin/pgcopydb/ld_transform.c

```diff
@@ -2609,3 +2653,269 @@ LogicalMessageValueEq(LogicalMessageValue *a, LogicalMessageValue *b)
+		return false;
+	}
+
+	(void) normalizedGeneratedColumn(&(item->generatedColumn),
+									 column->nspname,
+									 column->relname,
+									 column->attname);
+
+	HASH_ADD(hh,
+			 privateContext->generatedColumnsCache,
+			 generatedColumn,
+			 sizeof(GeneratedColumn),
+			 item);
+
+	return true;
+}
+
+
+/*
+ * prepareGeneratedColumnsCache fills-in the cache with the tables having
+ * generated columns.
+ */
+static bool
+prepareGeneratedColumnsCache(StreamSpecs *specs)
+{
+	/*
+	 * TODO: GeneratedColumn must be retrieved from the target catalog
+	 * because the schema of the target can be different from the source.
+	 */
+	if (!catalog_iter_s_generated_column(specs->sourceDB,
+										 specs,
+										 &generated_column_iter_hook))
+	{
+		log_error("Failed to prepare a generated column cache for our catalog,"
+				  "see above for details");
+		return false;
+	}
+
+	return true;
+}
+
+
+/*
+ * removeGeneratedColumnsFromTransaction removes the generated columns from the
+ * given transaction.
+ */
+static bool
+removeGeneratedColumnsFromTransaction(GeneratedColumnsCache *cache,
+									  LogicalTransaction *txn)
+{
+	LogicalTransactionStatement *stmt = txn->first;
+
+	for (; stmt != NULL; stmt = stmt->next)
+	{
+		if (!removeGeneratedColumnsFromStatement(cache, stmt))
+		{
+			return false;
+		}
+	}
+
+	return true;
+}
+
+
+/*
+ * removeGeneratedColumnsFromStatement removes the generated columns from the
+ * given statement after looking up the cache.
+ */
+static bool
+removeGeneratedColumnsFromStatement(GeneratedColumnsCache *cache,
+									LogicalTransactionStatement *stmt)
+{
+	LogicalMessageTupleArray *columns = NULL;
+	const char *nspname = NULL;
+	const char *relname = NULL;
+
+	if (stmt->action == STREAM_ACTION_INSERT)
+	{
+		columns = &(stmt->stmt.insert.new);
+		nspname = stmt->stmt.insert.table.nspname;
+		relname = stmt->stmt.insert.table.relname;
+	}
+	else if (stmt->action == STREAM_ACTION_UPDATE)
+	{
+		columns = &(stmt->stmt.update.new);
+		nspname = stmt->stmt.update.table.nspname;
+		relname = stmt->stmt.update.table.relname;
+	}
+	else
+	{
+		/* Only INSERT and UPDATE statements might have generated columns */
+		return true;
+	}
+
+	if (columns->count > 1)
+	{
+		log_error("BUG: removeGeneratedColumnsFromStatement called on a statement "
+				  "with multiple VALUES clauses");
+		return false;
+	}
+
+	LogicalMessageTuple *tuple = &(columns->array[0]);
+
+	for (int c = 0; c < tuple->cols; c++)
+	{
+		if (isGeneratedColumn(cache,
+							  nspname,
+							  relname,
+							  tuple->columns[c]))
+		{
+			/* deallocate the column */
+			free(tuple->columns[c]);
+
+			/* remove the column from the tuple */
+			for (int j = c; j < tuple->cols - 1; j++)
+			{
+				tuple->columns[j] = tuple->columns[j + 1];
+			}
+
+			/* number of colums have reduced now */
+			tuple->cols--;
+
+			/* remove the value from the tuple */
+			for (int r = 0; r < tuple->values.count; r++)
+			{
```

Comment:
We could keep the column in there and replace the value with the DEFAULT keyword, I believe. Does that work? does it behave better in terms of performance characteristics? Is it easier to debug?

Reply:
Yeah, it was one of the candidates. But, since we pass the EXEC parameters as a JSON string, we may not be able to distinguish between a DEFAULT keyword and DEFAULT as a string unless we make changes to the way we handle it.

Reply:
I think we can use a special marker object to encode DEFAULT keyword, something like below,

> EXECUTE 7fa747a0["20","2024-06-04 11:41:19.279866+00", {"keyword": "DEFAULT"}];

Reply:
I don't know if keeping the number of attributes is important enough to warrant that change. What I like about it is that we don't need to edit the columns and values arrays (run-time performance characteristics should be better), and it makes it easier to debug too. How does it look like in terms of complexity? do you think such a change is a good idea, and that it's worth it?

Reply:
@dimitri It has been simplified now and uses DEFAULT as a value instead of removing the column.

Reply:
Thanks for that extra work!